File: algorithms/dynamic-programming/word-break/cpp/can_sum.cpp
#include <iostream>
#include <vector>
using namespace std;

/**
 * Determine if target can be formed by summing elements from arr
 * with repetition allowed.
 *
 * arr: vector of positive integers (available elements)
 * target: the target sum to reach
 * Returns: 1 if target is achievable, 0 otherwise
 */
int canSum(const vector<int>& arr, int target) {
    if (target == 0) return 1;

    vector<bool> dp(target + 1, false);
    dp[0] = true;

    for (int i = 1; i <= target; i++) {
        for (int elem : arr) {
            if (elem <= i && dp[i - elem]) {
                dp[i] = true;
                break;
            }
        }
    }

    return dp[target] ? 1 : 0;
}

int main() {
    cout << canSum({2, 3}, 7) << endl;   // 1
    cout << canSum({5, 3}, 8) << endl;   // 1
    cout << canSum({2, 4}, 7) << endl;   // 0
    cout << canSum({1}, 5) << endl;      // 1
    cout << canSum({7}, 3) << endl;      // 0
    return 0;
}

File: algorithms/dynamic-programming/word-break/python/can_sum.py
def can_sum(arr, target):
    """
    Determine if target can be formed by summing elements from arr
    with repetition allowed.

    arr: list of positive integers (available elements)
    target: the target sum to reach
    Returns: 1 if target is achievable, 0 otherwise
    """
    if target == 0:
        return 1

    dp = [False] * (target + 1)
    dp[0] = True

    for i in range(1, target + 1):
        for elem in arr:
            if elem <= i and dp[i - elem]:
                dp[i] = True
                break

    return 1 if dp[target] else 0


if __name__ == "__main__":
    print(can_sum([2, 3], 7))   # 1 (2+2+3)
    print(can_sum([5, 3], 8))   # 1 (3+5)
    print(can_sum([2, 4], 7))   # 0
    print(can_sum([1], 5))      # 1 (1+1+1+1+1)
    print(can_sum([7], 3))      # 0
